MyJetpack: redirect after checkout to Jetpack recommendations page if plugin is active

Initsogar opened the pull request:
If we purchase a backup plan through Dashboard page in Jetpack plugin, it will redirect back to `admin.php?page=jetpack#/recommendations`, but if we purchase through My Jetpack, it will redirect back to My Jetpack.

We have a couple of issues created related to this interaction, involving Backup and Scan products. And after discussing in p6TEKc-6ZE-p2, we require to do the following changes in the interaction with Backup and Scan:

* If Jetpack plugin is active:
  * After purchasing a Backup or Scan plan, it should redirect to Jetpack recommendations page.
* If VaultPress Backup plugin is active:
  * After purchasing a Backup plan, it should redirect to backup plugin admin page.

Fixes #29506, fixes #29532 and fixes #29591.

## Proposed changes:

* Add new `post_checkout_url` field and `get_post_checkout_url` static method on MyJetpack Products class and implement it in Backup and Scan products to return the desired post checkout URL depending on Jetpack plugin availability.
* If Jetpack plugin is active, after purchasing Backup or Scan, customers will be redirected to `admin.php?page=jetpack#/recommendations`.
* If VaultPress Backup plugin is active and Jetpack is inactive, customers will be redirected to backup plugin admin page `admin.php?page=jetpack-backup`.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
* p1HpG7-l2l-p2
* p6TEKc-6ZE-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions - Backup

### Using Jetpack plugin
* Spin up a Jurassic Ninja site with Jetpack plugin
* Connect the site and user to Jetpack
* Navigate to Jetpack > My Jetpack
* Click on `Add VaultPress Backup` button under `Jetpack VaultPress Backup` product card
  <img src="https://user-images.githubusercontent.com/1488641/227659770-0186ed2d-1b30-4b2e-a775-69d43394dc1d.png" width="200" />
* Click on `Add Jetpack VaultPress Backup` button on the interstitial pricing page
  <img src="https://user-images.githubusercontent.com/1488641/227659840-7e9648ea-1284-40af-a51d-291a6a084fcd.png" width="200" />
* It should redirect you to the checkout page with the Jetpack VaultPress Backup 10 GB plan in the cart.
* Purchase the plan.
* It should redirect you back to `[site admin url]/admin.php?page=jetpack#/recommendations`. You should see this page:
  <img src="https://user-images.githubusercontent.com/1488641/227659958-4dc8b5c0-5d4f-44e9-a582-b0c672f740f9.png" width="200" />

#### Using Jetpack VaultPress Backup plugin and Jetpack plugin is inactive
* Spin up a Jurassic Ninja site with Jetpack VaultPress Backup plugin (Jetpack plugin should be inactive or not installed)
* Connect the site and user to Jetpack
* Navigate to Jetpack > My Jetpack
* Click on `Purchase` button under `Jetpack VaultPress Backup` product card
  <img src="https://user-images.githubusercontent.com/1488641/227660989-eacd8dd0-fada-47c3-a5a6-82ae1b3b19ef.png" width="200" />
* Click on `Add Jetpack VaultPress Backup` button on the interstitial pricing page
* It should redirect you to the checkout page with the Jetpack VaultPress Backup plan in the cart.
* Purchase the plan.
* Validate it should redirect you to the backup plugin admin page (/wp-admin/admin.php?page=jetpack-backup).

## Testing instructions - Scan
* Spin up a Jurassic Ninja site with Jetpack plugin
* Connect the site and user to Jetpack
* Navigate to Jetpack > My Jetpack
* Click on `Purchase` button under `Scan` product card
  <img src="https://user-images.githubusercontent.com/1488641/227668853-8f875f87-61e9-4c29-91e7-a3f821b68d25.png" width="200" />
* Click on `Add Jetpack Scan` button on the interstitial pricing page
* It should redirect you to the checkout page with the Jetpack Scan plan in the cart.
* Purchase the plan.
* It should redirect you back to `[site admin url]/admin.php?page=jetpack#/recommendations`. You should see this page:
  <img src="https://user-images.githubusercontent.com/1488641/227669334-56f99db0-02a9-4770-a416-de27941d4035.png" width="200" />

Initsogar committed:
Conditionally redirect to MyJetpack or Jetpack recommendations page after purchase

Initsogar committed:
changelog

Initsogar committed:
Changelog significance to patch instead of minor

Initsogar committed:
Pass checkoutRedirectUrl to trial checkout redirect

Initsogar committed:
Add isJetpackPluginActive to MyJetpack initial state

Initsogar committed:
Redirect trials to MyJetpack

Initsogar committed:
Validate if jetpack and jetpack-dev plugin is active

jwebbdev reviewed (dismissed):
I tested the 3 flows and all worked well.  Code seems reasonable, but wouldn't mind someone with more React-fu to chime in before we merge in case there's an existing method to check for Jetpack plugin being active here.

miguelxpn reviewed (dismissed):
Code looks good and tests great! Thanks for working on this!

retrofox commented on projects/packages/my-jetpack/src/class-initializer.php:
```diff
@@ -168,6 +168,7 @@ public static function enqueue_scripts() {
 				'fileSystemWriteAccess' => self::has_file_system_write_access(),
 				'loadAddLicenseScreen'  => self::is_licensing_ui_enabled(),
 				'adminUrl'              => esc_url( admin_url() ),
+				'isJetpackPluginActive' => Plugins_Installer::is_plugin_active( 'jetpack/jetpack.php' ) || Plugins_Installer::is_plugin_active( 'jetpack-dev/jetpack.php' ),
```
I don't know whether it should be handled here. Since if so, we'll probably add other plugins with a similar situation. Isn't it possible to do it straight in the Backup plugin instead?

Initsogar replied:
@retrofox  I noticed that we require the same behavior in backup, scan, and also in VideoPress according to this task: https://github.com/Automattic/jetpack/issues/29593, and all of them are returning to My Jetpack instead of going to the recommendations page.

Given that My Jetpack is the shared package among them, and is the one package doing it differently, we suggest including it here.

Unfortunately, I wasn't able to reproduce the VideoPress scenario because I didn't see the option to actually purchase from Jetpack, just to add/activate the plugin.

What do you think?

renatoagds replied:
Agree with this one. Maybe it is time to open support to customize the checkout URL redirect.

Suggestion for this one @Initsogar :

Open a new field and static method at the Product class (`checkoutUrl` maybe): https://github.com/Automattic/jetpack/blob/665cff874552e344f415461f4e5ed41914f6c6d4/projects/packages/my-jetpack/src/products/class-product.php#L109-L134

----

Implement it in the [Backup](https://github.com/Automattic/jetpack/blob/665cff874552e344f415461f4e5ed41914f6c6d4/projects/packages/my-jetpack/src/products/class-backup.php) class.

You could use this utility method to check for jetpack plugin availability: https://github.com/Automattic/jetpack/blob/665cff874552e344f415461f4e5ed41914f6c6d4/projects/packages/my-jetpack/src/products/class-product.php#L333-L335

----

Then you will have it available at the `detail` node https://github.com/Automattic/jetpack/blob/665cff874552e344f415461f4e5ed41914f6c6d4/projects/packages/my-jetpack/_inc/components/product-detail-card/index.jsx#L73-L84

And you can check for it and apply it if available or `myJetpackUrl` if not.

renatoagds replied:
I posted mine after yours @Initsogar ... But if we want to move everyone to `recommendations`, I think this is probably the way. Should we just confirm this with design?

retrofox replied:
Thanks, Renato 💯 

Also and if I'm not wrong, wondering whether leading the user to the Jetpack plugin instead of the standalone is something we want to do since it seems to go in the opposite direction to the vision of the standalone plugins' goal. Anyway, that's another topic :-)

Initsogar replied:
Thanks for the feedback @renatoagds and @retrofox. Valuable insights!

I just requested feedback from design to understand the upgrade flow approach through My Jetpack. Noticed that for Jetpack Search they are asking for something a bit different in https://github.com/Automattic/jetpack/issues/29592 so it would be great to clarify.

It would be great to clarify that before making any changes. Either way, your approach looks neat @renatoagds!

p6TEKc-6ZE-p2

renatoagds reviewed (dismissed):
Added some suggestion to make it, but agree that it should be more product related.

Initsogar committed:
[not verified] Merge remote-tracking branch 'origin/trunk' into update/myjetpack-checkout-redirect

Initsogar committed:
Add post_checkout_url to MyJetpack products

Initsogar committed:
Remove isJetpackPluginActive from MyJetpack state

Initsogar committed:
Update package version

Initsogar committed:
Add support for Scan

Initsogar committed:
Update changelog

Initsogar committed:
Return null if Jetpack is not active in Scan

Initsogar committed:
Ensure redirect to plugin admin page if active

Initsogar commented:
Hey there!

Resuming the work on this PR, and after getting feedback from design team on p6TEKc-6ZE-p2 it makes sense to pivot the initial idea and make it more product related as @renatoagds and @retrofox suggested.

We just added a new `get_post_checkout_url` in Products class and implemented it in Backup and Scan products classes to return the expected URL we want to redirect customers to after purchasing these plans, depending on whether Jetpack plugin is active. Thanks @renatoagds for the idea!

The expected behavior is that when purchasing Backup or Scan with Jetpack plugin active, it will redirect customers to our Jetpack recommendations page. In case we are purchasing Backup with the VaultPress Backup plugin active (and Jetpack inactive) it will redirect to backup plugin admin page according.

With this approach, we will solve the following open issues:
* https://github.com/Automattic/jetpack/issues/29506
* https://github.com/Automattic/jetpack/issues/29532
* https://github.com/Automattic/jetpack/issues/29591

It would be great if you could take another look :)

Thank you so much for your feedback!

lhkowalski commented:
Code change is neat and makes sense. Going to follow the testing instructions now.

lhkowalski commented:
Did the first test using only the Jetpack plugin and adding VaultPress. The call to action on the card changed a bit after moving it back to a hybrid product (#29906) (it says "Purchase" now), but that's expected since we are doing some fixes on the My Jetpack statuses.

I followed the checkout flow and got redirected to the right page after it:

<img width="1072" alt="Screen Shot 2023-04-13 at 11 58 28" src="https://user-images.githubusercontent.com/6760046/231801819-7570540b-6986-4493-b4cc-42fa2a850059.png">

So it's working as expected.

lhkowalski commented:
Did the second test using only the VaultPress backup standalone plugin:

<img width="259" alt="Screen Shot 2023-04-13 at 12 08 02" src="https://user-images.githubusercontent.com/6760046/231803509-adf3fd29-8e02-4422-9d71-eab81a875c3d.png">

I clicked "Purchase" and got to the interstitial. Some recent changes changed the wording there as well, so now instead of `Add Jetpack VaultPress Backup` it says `Get Jetpack VaultPress Backup`.

I finished the checkout flow and got redirected to `/wp-admin/admin.php?page=jetpack-backup`:

<img width="1101" alt="Screen Shot 2023-04-13 at 12 12 10" src="https://user-images.githubusercontent.com/6760046/231804700-a0129fd2-dd9e-48be-9d02-be4710da651d.png">

So it's also working as expected.

lhkowalski commented:
Did the third test for Scan, using only the Jetpack plugin:

<img width="213" alt="Screen Shot 2023-04-13 at 12 13 38" src="https://user-images.githubusercontent.com/6760046/231805121-5b1e5cbd-5186-4c48-bcb0-008c3831fbec.png">

I clicked the "Purchase" button and got to the interstitial pricing page. The wording is also updated there, so we now see "Get Jetpack Scan".

I followed the checkout flow and landed on expected screen:

<img width="1067" alt="Screen Shot 2023-04-13 at 12 20 11" src="https://user-images.githubusercontent.com/6760046/231806952-fc1f67cc-c914-4ca1-b440-21f3304ff1e8.png">

Again, working as expected.

lhkowalski reviewed (approved):
The changes are testing as expected and the code is neat and clear. I see this as a good improvement because now we can customize the post-checkout URL for every product and in every context (Jetpack or not, standalone or not). Nice work!
